policies/templates/gcp_storage_bucket_retention_v1.yaml

```diff
@@ -0,0 +1,212 @@
+        # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+        # See the License for the specific language governing permissions and
+        # limitations under the License.
+        #
+        
+        package templates.gcp.GCPStorageBucketRetentionConstraintV1
+        
+        import data.validator.gcp.lib as lib
+        
+        deny[{
+        	"msg": message,
+        	"details": metadata,
+        }] {
+        	constraint := input.constraint
+        	lib.get_constraint_params(constraint, params)
+        
+        	min_retention_days := lib.get_default(params, "minimum_retention_days", "")
+        	max_retention_days := lib.get_default(params, "maximum_retention_days", "")
+        
+        	asset := input.asset
+        	asset.asset_type == "storage.googleapis.com/Bucket"
+        
+        	# Check if resource is in exempt list
+        	exempt_list := params.exemptions
+        	matches := {asset.name} & cast_set(exempt_list)
+        	count(matches) == 0
+        
+        	violation_msg := get_diff(asset, min_retention_days, max_retention_days)
+        	is_string(violation_msg)
+        
+        	message := sprintf("Storage bucket %v has a retention policy violation: %v", [asset.name, violation_msg])
+        
+        	metadata := {
+        		"resource": asset.name,
+        		"violation_type": violation_msg,
+        	}
+        }
+        
+        ###########################
+        # Rule Utilities
+        ###########################
+        
+        # Generate a violation if there is no bucket lifecycle Delete condition and maximum_retention_days is defined.
+        get_diff(asset, minimum_retention_days, maximum_retention_days) = "Lifecycle delete action does not exist when maximum_retention_days is defined" {
+        	maximum_retention_days != ""
+        	lifecycle_delete_not_exists := [is_not_delete | is_not_delete = asset.resource.data.lifecycle.rule[_].action.type != "Delete"]
+        	all(lifecycle_delete_not_exists)
+        }
+        
+        # Generate a violation if the bucket lifecycle Delete 'age' condition is greater than the maximum_retention_days defined.
+        get_diff(asset, minimum_retention_days, maximum_retention_days) = "Lifecycle age is greater than maximum_retention_days" {
+        	some i
+        	maximum_retention_days != ""
+        	asset.resource.data.lifecycle.rule[i].action.type == "Delete"
+        	lifecycle_age := asset.resource.data.lifecycle.rule[i].condition.age
+        	lifecycle_age > maximum_retention_days
+        }
+        
+        # Generate a violation if the bucket lifecycle Delete 'age' condition does NOT exist and maximum_retention_days is defined.
+        get_diff(asset, minimum_retention_days, maximum_retention_days) = "Lifecycle age is not set when maximum_retention_days is defined" {
+        	some i
+        	maximum_retention_days != ""
+        	asset.resource.data.lifecycle.rule[i].action.type == "Delete"
```

Comment:
Yes it is possible, but since this is just looking at the maximum_retention_days (the existence of a maximum number of days a resource should exist), if the age of a bucket does not exist, there should be a violation.